File: bin/diff_bams.pl
#!/usr/bin/perl

##########LICENCE##########
# PCAP - NGS reference implementations and helper code for the ICGC/TCGA Pan-Cancer Analysis Project
# Copyright (C) 2014 ICGC PanCancer Project
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not see:
#   http://www.gnu.org/licenses/gpl-2.0.html
##########LICENCE##########

use strict;
use autodie qw(:all);
use warnings FATAL => 'all';
use Getopt::Long;
use Pod::Usage qw(pod2usage);
use PCAP::Cli;
use FindBin qw($Bin);
use lib "$Bin/../lib";

use Bio::DB::HTS;
use Bio::DB::HTS::AlignWrapper;

my ($file_a, $file_b, $skip_z, $count_flag_diff) = setup();

my $sam_a = Bio::DB::HTS->new(-bam => $file_a);
my $bam_a = $sam_a->hts_file;
my $bam_b = Bio::DB::HTSfile->open($file_b);

my $header_a       = $bam_a->header_read;
my $target_count_a = $header_a->n_targets;
my $target_names_a = $header_a->target_name;

my $header_b       = $bam_b->header_read;
my $target_count_b = $header_b->n_targets;
my $target_names_b = $header_b->target_name;

unless($target_count_a == $target_count_b) {
  die "Reference sequence count is different\n";
}
print "Reference sequence count passed\n";

for(0..($target_count_a-1)) {
  die "Reference sequences in different order\n" if($target_names_a->[$_] ne $target_names_b->[$_]);
}
print "Reference sequence order passed\n";

my ($align_a, $align_b, $count);
my $flag_diffs = 0;
$|++;
my $last_coord = 0;
my %dup_pileup;
while(1) {
  $count++;
  $align_a = $bam_a->read1($header_a);
  $align_b = $bam_b->read1($header_b);
  if($skip_z) {
    while($align_a->qual == 0) { $align_a = $bam_a->read1($header_a); last unless(defined $align_a); }
    while($align_b->qual == 0) { $align_b = $bam_b->read1($header_b); last unless(defined $align_b); }
  }

  last if(!defined $align_a && !defined $align_b);
  die "Files have different number of records\n" if((!defined $align_a && defined $align_b) || (defined $align_a && !defined $align_b));

  die sprintf "Files differ at record $count (qname) a=%s b=%s\n", $align_a->qname, $align_b->qname if($align_a->qname ne $align_b->qname);
  if($align_a->flag ne $align_b->flag) {
    if($count_flag_diff) {
      $flag_diffs++;
      my $pos = $align_a->pos;
      if(($pos - $last_coord) >= 0) { # saves checking for different chr
        my $aw = Bio::DB::HTS::AlignWrapper->new($align_a, $sam_a);
        $dup_pileup{$aw->seq_id}{$pos} += 1;
      }
      $last_coord = $pos;
    }
    else {
      die sprintf "Files differ at record $count (flags) a=%s b=%s\n", $align_a->qname, $align_b->qname
    }
  }

  if($count % 1_000_000 == 0) {
    my $message = "Matching records: $count";
    $message .= "\t(flag mismatch: $flag_diffs)"if($count_flag_diff);
    print $message."\r";
  }
}
print "Matching records: $count\n";
if($count_flag_diff) {
  print "Flag mismatches: $flag_diffs\n";
  print "Locations of flag differences:\n";
  print "#Chr\tPos\tCount\n";
  for my $chr(sort keys %dup_pileup) {
    for my $pos(sort {$a<=>$b} keys %{$dup_pileup{$chr}}){
      print sprintf "%s\t%s\t%s\n", $chr, $pos, $dup_pileup{$chr}{$pos};
    }
  }
}

sub setup {
  my %opts;
  GetOptions( 'h|help' => \$opts{'h'},
              'm|man' => \$opts{'m'},
              'a|bam_a=s' => \$opts{'a'},
              'b|bam_b=s' => \$opts{'b'},
              'r|ref=s' => \$opts{'r'},
              'c|count' => \$opts{'c'},
              's|skip' => \$opts{'s'},
  ) or pod2usage(2);

  pod2usage(-message => PCAP::license, -verbose => 1) if(defined $opts{'h'});
  pod2usage(-message => PCAP::license, -verbose => 2) if(defined $opts{'m'});

  my $defined = 0;
  for(keys %opts) { $defined++ if(defined $opts{$_}); }
  if($defined == 0 && @ARGV == 2) {
    $opts{'a'} = shift @ARGV;
    $opts{'b'} = shift @ARGV;
    $defined = 2;
  }
  pod2usage(-msg  => "\nERROR: Options 'a' & 'b' must be defined.\n", -verbose => 1,  -output => \*STDERR) unless($defined >= 2);

  die pod2usage(-msg  => "\nERROR: bam_a and bam_b cannot be the same file\n", -verbose => 1,  -output => \*STDERR) if($opts{'a'} eq $opts{'b'});

  PCAP::Cli::file_for_reading('bam_a', $opts{'a'});
  PCAP::Cli::file_for_reading('bam_a', $opts{'b'});
  if($opts{'a'} =~ m/[.]cram/ || $opts{'b'} =~ m/[.]cram/) {
    die pod2usage(-msg  => "\nERROR: Option '-r' must be defined if any CRAM files are provided\n", -verbose => 1,  -output => \*STDERR) unless(defined $opts{'r'});
    PCAP::Cli::file_for_reading('ref', $opts{'r'});
  }

  return ($opts{'a'}, $opts{'b'}, $opts{'s'}, $opts{'c'});
}

__END__

=head1 NAME

diff_bams.pl - Compares two BAM/CRAM files irrespective of irrelevant header file differences.

=head1 SYNOPSIS

diff_bams.pl 1.bam 2.bam
  or
diff_bams.pl -a 1.bam -b 2.bam

It is rare that BAM files can be compared via MD5 due to various processing information being
incorporated into the header such as input/output paths.  This program works around this by only
comparing relevant information such as:

  1. Same number and order of @SQ headers.
  2. All reads are in same order.

You are able to compare a BAM vs it's corresponding CRAM

  Required parameters:
    -bam_a    -a    The first BAM|CRAM file.
    -bam_b    -b    The second BAM|CRAM file.

  Other:
    -ref      -r    Required for CRAM, genome.fa with co-located fai.
    -count    -c    Count flag differences
    -skipz    -s    Don't include reads with MAPQ=0 in comparison
    -help     -h    Brief help message.
    -man      -m    Full documentation.

  Run with '-m' for more detail

=head1 OPTIONS

=over 8

=item B<-bam_a>

A valid readable BAM file

=item B<-bam_b>

A valid readable BAM file (different to bam_a)

=item B<-skipz>

As reads with MAPQ zero can end up in multiple locations don't consider them
as a mismatch.

=item B<-help>

Print a brief help message and exits.

=item B<-man>

Prints the manual page and exits.

=back

File: c/diff_bams.c
#include <getopt.h>
#include <inttypes.h>
#include "htslib/sam.h"
#include "khash.h"
#include "dbg.h"

KHASH_MAP_INIT_INT(posn,int32_t)
KHASH_MAP_INIT_INT(chrom,khash_t(posn))

char *bam_a_loc = NULL;
char *bam_b_loc = NULL;
char *ref_file = NULL;
int skip_z = 0;
int count_flag_diff = 0;


int check_exist(char *fname){
	FILE *fp;
	if((fp = fopen(fname,"r"))){
		fclose(fp);
		return 1;
	}
	return 0;
}

void print_version (int exit_code){
  printf ("%s\n",VERSION);
	exit(exit_code);
}

void print_usage (int exit_code){

	printf ("Usage: diff_bams -a bam_a.bam -b bam_b.bsm [-r reference.fa] [] [] [-h] [-v]\n\n");
	printf ("Required:\n");
	printf ("-a --bam_a          The first BAM|CRAM file.\n");
	printf ("-b --bam_b          The second BAM|CRAM file.\n\n");
	printf ("Other:\n");
  printf ("-r --ref            Required for CRAM, genome.fa with co-located fai.\n");
  printf ("-c --count          Count flag differences.\n");
  printf ("-s --skip           Don't include reads with MAPQ=0 in comparison.\n\n");
  printf ("-h --help           Display this usage information.\n");
	printf ("-v --version        Prints the version number.\n\n");
  exit(exit_code);
}

void options(int argc, char *argv[]){

  const struct option long_opts[] =
    {
              {"version", no_argument, 0, 'v'},
              {"help",no_argument,0,'h'},
              {"bam_a",required_argument,0,'a'},
              {"ref",required_argument,0,'r'},
              {"bam_b",required_argument,0,'b'},
              {"skip",no_argument,0,'s'},
              {"count",no_argument,0,'c'},
              { NULL, 0, NULL, 0}

   }; //End of declaring opts

   int index = 0;
   int iarg = 0;

     //Iterate through options
   while((iarg = getopt_long(argc, argv, "a:b:r:scvh", long_opts, &index)) != -1){
    switch(iarg){
      case 's':
        skip_z = 1;
        break;

      case 'c':
        count_flag_diff = 1;
        break;

      case 'r':
        ref_file = optarg;
        break;

      case 'a':
        bam_a_loc = optarg;
        break;

      case 'b':
        bam_b_loc = optarg;
        break;

      case 'h':
        print_usage(0);
        break;

      case 'v':
        print_version(0);
        break;

      case '?':
        print_usage (1);
        break;

      default:
        print_usage (1);

    }; // End of args switch statement

   }//End of iteration through options

   //Do some checking to ensure required arguments were passed and are accessible files
  if(ref_file != NULL){
    if(check_exist(ref_file) != 1){
      fprintf(stderr,"Reference fasta file (-r) %s does not exist.\n",ref_file);
      print_usage(1);
    }
  }

  if(strcmp(bam_a_loc,bam_b_loc)==0){
    fprintf(stderr,"bam_a '%s' and bam_b '%s' cannot be the same file\n",bam_a_loc,bam_b_loc);
    print_usage(1);
  }

  if(check_exist(bam_a_loc) != 1){
    fprintf(stderr,"First BAM|CRAM file (-a) %s does not exist.\n",bam_a_loc);
    print_usage(1);
  }

  if(check_exist(bam_b_loc) != 1){
    fprintf(stderr,"Second BAM|CRAM file (-b) %s does not exist.\n",bam_b_loc);
    print_usage(1);
  }

  if((strcmp(bam_a_loc + strlen(bam_a_loc) - 4,".cram")==0 || strcmp(bam_b_loc + strlen(bam_b_loc) - 4,".cram")==0) && ref_file == NULL) {
    fprintf(stderr,"Option '-r' must be defined if any CRAM files are provided\n");
    print_usage(1);
  }

   return;
}

int main(int argc, char *argv[]){
  htsFile *htsa = NULL;
  htsFile *htsb = NULL;
  bam_hdr_t *heada = NULL;
  bam_hdr_t *headb = NULL;
  khash_t(chrom) *chr_hash = NULL;
  bam1_t *reada = NULL;
  bam1_t *readb = NULL;
  options(argc, argv);
  //Open bam file a
  htsa = hts_open(bam_a_loc,"r");
  check(htsa != NULL, "Error opening hts file 'a' for reading '%s'.",bam_a_loc);
  //Open bam file b
  htsb = hts_open(bam_b_loc,"r");
  check(htsb != NULL, "Error opening hts file 'b' for reading '%s'.",bam_b_loc);

  if(ref_file){
    hts_set_fai_filename(htsa, ref_file);
    hts_set_fai_filename(htsb, ref_file);
  }

  heada = sam_hdr_read(htsa);
  check(heada != NULL, "Error reading header from opened hts file 'a' '%s'.",bam_a_loc);
  headb = sam_hdr_read(htsb);
  check(headb != NULL, "Error reading header from opened hts file 'b' '%s'.",bam_b_loc);

  if(heada->n_targets != headb->n_targets ){
    sentinel("Reference sequence count is different\n");
  }
  fprintf(stdout,"Reference sequence count passed\n");

  int i=0;
  for(i=0;i<heada->n_targets;i++){
    if(strcmp(heada->target_name[i],headb->target_name[i])!=0){
      sentinel("Reference sequences in different order\n");
    }
  }
  fprintf(stdout,"Reference sequence order passed\n");

  uint64_t count = 0;
  uint64_t flag_diffs = 0;
  uint64_t last_coord = 0;
  int chka = 0;
  int chkb = 0;
  reada = bam_init1();
  readb = bam_init1();
  while(1){
    count++;
    //Check the individual reads
    chka = sam_read1(htsa,heada,reada);
    chkb = sam_read1(htsb,headb,readb);

    if(skip_z==1){
      //Move to next non 0 MQ read
      while(reada->core.qual == 0 && chka >= 0) {
        chka = sam_read1(htsa,heada,reada);
      }
      while(readb->core.qual == 0 && chkb >= 0) {
        chkb = sam_read1(htsb,headb,readb);
      }
    }

    if(chka<0 && chkb<0){
      break;
    }
    if((chka>=0 && chkb <0) || (chkb>=0 && chka<0)){
      sentinel("Files have different number of records\n");
    }

    if(reada->core.tid != readb->core.tid || reada->core.pos != readb->core.pos || strcmp(bam_get_qname(reada),bam_get_qname(readb))!=0){
      sentinel("Files differ at record %"PRIu64" (qname) a=%s b=%s\n",count,bam_get_qname(reada),bam_get_qname(readb));
    }

    if(reada->core.flag != readb->core.flag){
      if(count_flag_diff==1){
        if(chr_hash == NULL){ chr_hash = kh_init(chrom);}
        flag_diffs++;
        int32_t pos = reada->core.pos;
        if((pos - last_coord) >=0 ){ //Checking for different chr
          int res;
          khiter_t k = kh_get(chrom, chr_hash, reada->core.tid); // query the hash table
          khash_t(posn) *pos_hash = NULL;
          if(k == kh_end(chr_hash)){
            k = kh_put(chrom,chr_hash,reada->core.tid,&res);
            pos_hash = kh_init(posn);
            khiter_t kpos = kh_put(posn,pos_hash,pos,&res);
            kh_value(pos_hash,kpos) = 1;
            kh_value(chr_hash,k) = *pos_hash;
          }else{
            *pos_hash = kh_value(chr_hash,k);
            khiter_t kpos = kh_get(posn, pos_hash, pos); // query the pos hash table
            if(kpos == kh_end(pos_hash)){
              kpos = kh_put(posn,pos_hash,pos,&res);
              kh_value(pos_hash,kpos) = 1;
            }else{
              kh_value(pos_hash,kpos) = kh_value(pos_hash,kpos)+1;
            }
            kh_value(chr_hash,k) = *pos_hash;
          }
        }
        last_coord = pos;
      }else{
        sentinel("Files differ at record %"PRIu64" (flags) a=%s b=%s\n",count,bam_get_qname(reada),bam_get_qname(readb));
      }
    }//End of if flags don't match
    if(count % 5000000 == 0) {
      fprintf(stdout,"Matching records: %"PRIu64"",count);
      if(count_flag_diff){
        fprintf(stdout,"\t(flag mismatch: %"PRIu64")",flag_diffs);
      }
      fprintf(stdout,"\r");
    }
  }//End of looping through all reads

  fprintf(stdout,"Matching records: %"PRIu64"\n",count);
  if(count_flag_diff && chr_hash != NULL){
    fprintf(stdout,"Flag mismatches: %"PRIu64"\n",flag_diffs);
    fprintf(stdout,"Locations of flag differences:\n");
    fprintf(stdout,"#Chr\tPos\tCount\n");
    khiter_t k;
    for (k = kh_begin(chr_hash); k != kh_end(chr_hash); ++k) {
      if (kh_exist(chr_hash, k)) {
        const int32_t keyp = kh_key(chr_hash,k);
        khash_t(posn) pos_hash = kh_value(chr_hash, k);
        khiter_t kpos;
        for (kpos = kh_begin(&pos_hash); kpos != kh_end(&pos_hash); ++kpos) {
          if (kh_exist(&pos_hash, kpos)) {
            const int32_t key = kh_key(&pos_hash,kpos);
            int32_t counter = kh_value(&pos_hash,kpos);
            fprintf(stdout,"%s\t%"PRIi32"\t%"PRIi32"\n",heada->target_name[keyp], key, counter);
            kh_destroy(posn,&pos_hash);
          }
        }

      }
    }
    kh_destroy(chrom,chr_hash);
  }

  bam_destroy1(reada);
  bam_destroy1(readb);
  bam_hdr_destroy(heada);
  bam_hdr_destroy(headb);
  hts_close(htsa);
  hts_close(htsb);
  return 0;
error:
  if(count_flag_diff && chr_hash != NULL){
    khiter_t k;
    for (k = kh_begin(chr_hash); k != kh_end(chr_hash); ++k) {
      if (kh_exist(chr_hash, k)) {
        const int32_t keyp = kh_key(chr_hash,k);
        khash_t(posn) pos_hash = kh_value(chr_hash, k);
        kh_destroy(posn,&pos_hash);
      }
    }
    kh_destroy(chrom,chr_hash);
  }
  if(reada) bam_destroy1(reada);
  if(readb) bam_destroy1(readb);
  if(heada) bam_hdr_destroy(heada);
  if(headb) bam_hdr_destroy(headb);
  if(htsa) hts_close(htsa);
  if(htsb) hts_close(htsb);
  return 1;
}